### PR TITLE
Support IO-like objects efficiently

### PR DESCRIPTION
**Pull request by janko-m**
[Shrine](https://github.com/janko-m/shrine) works with the abstraction of IO-like objects. For an object to be "IO-like" it proved to be enough that it responds to five IO methods: `#read`, `#size`, `#rewind`, `#eof?`, and `#close`.

However, MimeMagic expects the IO object to also respond to `#seek` (and also `#binmode`). If it doesn't respond to `#seek`, MimeMagic will currently read the whole file into memory. This will cause problems with large files and [remote files](https://github.com/janko-m/down#streaming).

Fortunately, the functionality of `io.seek(n)` can actually be produced using only the five methods mentioned above:

```rb
io.rewind
io.read(n)
```

This PR switches MimeMagic to use only these methods, allowing more generic IO-like objects.

### CODE REVIEW

File: test/mimemagic_test.rb [hunk 1]
```diff
@@ -113,20 +113,4 @@
...
-      end
-    end
-    MimeMagic.by_magic(ReadableObj.new).should.equal 'application/mimemagic-test'
-    class StringableObject
-      def to_s
-        'MAGICTEST'
-      end
-    end
-    MimeMagic.by_magic(StringableObject.new).should.equal 'application/mimemagic-test'
-  end
```
Comment (minad) [inline test/mimemagic_test.rb]: Why did you remove that test?

Comment (janko-m) [inline test/mimemagic_test.rb]: You're right, there was no reason, I added them back, but changed `ReadableObj` to `IOObject`. Even though MimeMagic currently uses only `#read` and `#rewind`, it might use other IO methods in the future, and I think it's good to set up this "IO contract" upfront.

**Comment by janko-m** [comment]
@minad Do you think we can merge this and release a new version? I would like to be able to have this available for Shrine users 😃

**Comment by minad** [comment]
Hi, sure go forward with it. I added you as a collaborator!

**Comment by janko-m** [comment]
Just wanted to check with you. Awesome! Do you think you could release a new version? I think it is still a patch version, as nothing new was added, just performance improvements.

**Comment by minad** [comment]
done :)

**Comment by janko-m** [comment]
Thanks!
